Sources/Plasma/PubUtilLib/plAvatar/plPhysicalControllerCore.cpp

```diff
@@ -375,194 +375,110 @@ plWalkingStrategy::plWalkingStrategy(plAGApplicator* rootApp, plPhysicalControll
-    if (ground.Normal.fZ < GetFallStopThreshold())
-        return;
 
-    // We are on the ground - good. However, there is a problem. Some ground colliders are
-    // built sloppily such that they intersect at (near) perpendicular angles and are just
-    // slightly too tall for the simulation to handle. So, if the actor did not move much this
-    // frame, we want to offset the character by some "step height" and sweep along the desired
-    // movement path. If we find that the shortest hit is a nontrival displacement, we should
-    // step up and move that far.
-    const hsVector3& animVelocity = fController->GetLinearVelocity();
-    const hsVector3& simVelocity = fController->GetAchievedLinearVelocity();
-    if (animVelocity.MagnitudeSquared() < 0.0001f || simVelocity.MagnitudeSquared() >= 0.0001f)
-        return;
+    fController->SetPushingPhysical(nullptr);
+    fController->SetFacingPushingPhysical(false);
+    fContacts.clear();
+    fFlags &= ~kResetMask;
 
-    hsPoint3 startPos;
-    fController->GetPositionSim(startPos);
-    hsPoint3 endPos = startPos + (animVelocity * delSecs * 2.5f);
-    startPos.fZ += fController->GetRadius() * 1.2f;
-
-    hsVector3 dir = (hsVector3)(endPos - startPos);
-    float delta = dir.Magnitude();
-    dir.Normalize();
-
-    // We don't want to step up onto kickables!
-    uint32_t simGroups = (1 << plSimDefs::kGroupStatic) | (1 << plSimDefs::kGroupAvatarBlocker);
-    auto newGround = fController->SweepSingle(startPos, dir, delta, (plSimDefs::Group)simGroups);
-
-    // Only step up onto the new ground if it's below our steepness limit
-    if (newGround && newGround->Displacement >= 0.001f &&
-        newGround->Normal.fZ >= GetFallStartThreshold() &&
-        newGround->Normal.fZ > 0.2f) {
-        hsVector3 displacement = dir * newGround->Displacement;
-        fController->SetPositionSim(startPos + displacement);
-        fController->OverrideAchievedLinearVelocity(displacement / delSecs);
-    }
+    uint32_t simGroups = (1 << plSimDefs::kGroupStatic) |
+                         (1 << plSimDefs::kGroupDynamic) |
+                         (1 << plSimDefs::kGroupAvatarBlocker);
+    if (!fController->IsSeeking())
+        simGroups |= (1 << plSimDefs::kGroupExcludeRegion);
+    uint32_t result = fController->Move(velocity, delSecs, (plSimDefs::Group)simGroups);
+    if (result & plPhysicalControllerCore::kBottom)
+        fFlags |= kGroundContact;
 }
 
 void plWalkingStrategy::Update(float delSecs)
 {
-    fFlags &= ~kResetMask;
-
-    if (auto ground = IFindGroundCandidate()) {
-        fFlags |= kGroundContact | kHitGroundInThisAge;
-        ICheckGroundSteepness(ground.value());
-        if (!IsControlledFlight())
-            IStepUp(ground.value(), delSecs);
-    }
-
-    if ((fFlags & kGroundContact) && !(fFlags & kFallingNormal)) {
+    if (fFlags & kGroundContact) {
         fTimeInAir = 0.f;
+        fFlags |= kHitGroundInThisAge;
```

Comment:
This flag is checked by some *very* unrelated code in the avatar modifier itself. I've really never traced that part down, but I know if you never tell that code, "well, I've hit the ground at least once," then the avatar can't be moved at all.

The flag is set here now because the way that we figure out ground contact is a bit different. Previously, I was examining the list of contact points and looking for something that appeared to be relatively flat on the bottom surface of the capsule. This needed to happen after the physics simulation, so that happened here in `Update()`. However, the PhysX CCT natively tells us if we have a contact below the character in the main `Apply()` phase, so the ground contact is now set there. I kept the `kHitGroundInThisAge` bit in `Update()` because of the seemingly related by topic reset of `fTimeInAir`.